Add smoketest to e2e tests

piersy opened the pull request:
Adds an e2e smoketest that sends value transfer, contract interaction
and contract creation transactions for all of the valid transaction types.

It also verifies that deprecated transactions (celo legacy & cip42) are
not supported.

piersy committed:
Add smoketest to e2e tests

The smoketest tries all types of transaction for both sends and creates.

karlb commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,170 @@
+const testContractJSON = JSON.parse(fs.readFileSync(process.env.COMPILED_TEST_CONTRACT, 'utf8'));
+
+
+// Setup up chain
+const devChain = defineChain({
+	...celoAlfajores,
+	id: 1337,
+	name: "local dev chain",
+	network: "dev",
+	rpcUrls: {
+		default: {
+			http: [process.env.ETH_RPC_URL],
+		},
+	},
+});
+
+const chain = (() => {
+	switch (process.env.NETWORK) {
+		case 'alfajores':
+			return celoAlfajores
+		default:
+			return devChain
+	};
+})();
```
We could start moving that to a shared file and maybe even include the `publicClient` and `walletClient ` below. It is already used in both `test_viem_tx.mjs` and `send_tx.mjs`.

piersy replied:
Good point, I was thinking about this.

ezdac replied:
I vote for that as well.
You could define some `setup.js` library file that takes in some parameters (`chain-id`, `private-key`, optional overwrite `rpc-url`) and returns the public and private viem client instances.
The alfajores chain switch could be activated by using its chain-id in the parameters.

piersy replied:
Ok I've made this change.

karlb commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,170 @@
+	it("send legacy tx", async () => {
+		const type = "legacy";
+		await verifyTypedTransactions(type);
+	}).timeout(30_000);
```
Is there a reason why this timeout is larger than the others?

piersy replied:
No, updated!

karlb commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,170 @@
+		await verifyTypedTransactions(type, process.env.FEE_CURRENCY);
+	}).timeout(10_000);
+
+	// This test is failing because the produced transaction is of type cip64.
+	// I guess this is a problem with the viem internals.
+	it.skip("cip42 not supported", async () => {
+		const type = "cip42";
+		await verifyTypedTransactions(type);
+	}).timeout(10_000);
```
Interesting and I think your explanation is correct. We could try running the same test against Celo L1 to make sure it's not a change in the chain behaviour that is causing this.

ezdac replied:
@piersy just to get the explanation right - viem with the new l2 alfajores definitions does not seem to know about "cip42" types and instead produces a cip64 transaction, that then fails the assertion? 

I would classify this as a viem bug, since it shouldn't even allow the deprecated `cip42' type and fail early on unknown types (at least in my expectation).

piersy replied:
yeah, I'd classify it as a viem bug, i will raise it with the dev tooling team

piersy replied:
However viem still needs to be able to send cip42 txs to the celo l1, so it does still need that functionality.

karlb replied:
Does it? It could just send CIP-64 without missing any features.

piersy replied:
Unless people really want to set the gatewayFee and gatewayFeeRecipient! I'll check with the devtooling team to see what the expected approach for viem is.

karlb replied:
You can't set those to non-zero values anymore (celo-blockchain will reject the tx otherwise), so nothing is lost.

piersy replied:
Ok, was unaware!

piersy replied:
@karlb @ezdac I've moved this test into the smoketest_unsupported_txs golang test. Where we can be sure that we are sending an invalid transaction, and therefore ensure that the server correctly rejects it.

karlb commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,170 @@
+			// transaction was not included.
+			let blockNumber = await publicClient.getBlockNumber();
+			const oldBlockNumber = blockNumber;
+			while(blockNumber == oldBlockNumber){
+				blockNumber = await publicClient.getBlockNumber();
+			}
```
A small sleep (e.g. 50ms) would be prevent people with low latency connections (or a script in the same datacenter) to send a large amount of requests while waiting. `getBlockNumber` should be fast enough for that to not matter too much, but let's avoid being spammy on principle.

karlb commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,170 @@
+			assert.fail("Managed to send unsupported legacy tx with fee currency");
+	}).timeout(10_000);
+
+	it.only("legacy create tx with fee currency not supported", async () => {
```
I'm not sure if it's worth having this test in addition to the previous one. Chances that tx types behave totally differently between send and create are low and we already test plenty of both in other tests with having to repeat them.

piersy replied:
Hey @karlb  I've removed these tests from the viem test and added them in golang since its easier to construct known unsupported transactions in golang, in the golang version I only test send transactions.

karlb commented on e2e_test/test_smoketest.sh:
```diff
@@ -0,0 +1,9 @@
+
+(cd debug-fee-currency && forge build .)
+export COMPILED_TEST_CONTRACT=../debug-fee-currency/out/DebugFeeCurrency.sol/DebugFeeCurrency.json
+cd js-tests && ./node_modules/mocha/bin/mocha.js test_viem_smoketest.mjs --exit
```
We could add a default timeout of 10 seconds here (`--timeout 10000`) if we don't prefer the explicitness of having it on every test.

piersy replied:
Nice, thanks!

piersy committed:
Fix erroneous timeout

piersy committed:
Remove use of only (left over from local debugging)

piersy committed:
Add a sleep to prevent spamming the server

piersy committed:
Set smoketest timeout globally

Chose a timeout of 16 seconds since the first two tests were taking 8
seconds on Alfajores.

karlb commented on e2e_test/test_smoketest.sh:
```diff
@@ -6,4 +6,4 @@ prepare_node
 (cd debug-fee-currency && forge build .)
 export COMPILED_TEST_CONTRACT=../debug-fee-currency/out/DebugFeeCurrency.sol/DebugFeeCurrency.json
-cd js-tests && ./node_modules/mocha/bin/mocha.js test_viem_smoketest.mjs --exit
+cd js-tests && ./node_modules/mocha/bin/mocha.js test_viem_smoketest.mjs --timeout 16000 --exit 
```
Here's a useless trailing space
```suggestion
cd js-tests && ./node_modules/mocha/bin/mocha.js test_viem_smoketest.mjs --timeout 16000 --exit
```

piersy replied:
What tools did you use to spot that?

karlb replied:
Just the github diff view.

ezdac commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,170 @@
+	const receipt = await publicClient.waitForTransactionReceipt({ hash: txHash });
+	const transaction = await publicClient.getTransaction({ hash: txHash });
+	assert.equal(transaction.type, type, "transaction type does not match");
+	assert.equal(receipt.status, "success", "receipt status 'failure'");
```
nitpick: I would put the assert conditions directly under the query in order to avoid doing the other query unnecessarily or potentially raising other, less specific errors during the other call.

ezdac commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,170 @@
+		await check(create, type);
+}
+
+describe("viem smoke test", () => {
```
Wouldn't it make sense to write one parameterized test and specify all types as parameter?
Also then it would be feasible to create separate tests for the create/non-create tx to give more failure context.

It seems mocha doesn't have a parameterization feature per se, but you could just iterate, something like:

```js
["legacy", "cip64"].forEach(function (type) {
    describe("Test tx type " + type, function () {
        it("send tx", function (done) {
          // sendTyped  etc.
        });
        it("send create tx", function (done) {
        // sendTypedCreate etc.
        });
    });
});
```

piersy replied:
yep, this looks nice, thanks!

piersy replied:
done

ezdac commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,170 @@
+// verifyTypedTransactions is a helper function that submits a send and create
+// transaction of the given type with optional feeCurrency and checks the
+// results.
+async function verifyTypedTransactions(type, feeCurrency){
```
I would remove this helper and write separate tests, see comment below

piersy replied:
done as per comment below

ezdac commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,170 @@
+			const hash = await sendTypedTransaction(type,process.env.FEE_CURRENCY);
+
+			// When using the devnet an exception is thrown from
+			// sendTypedTransaction, on alfajores the transaction is submitted so we
```
Similar to the comment with the skipped test above - shouldn't this be caught early by viem and refuse to send unsupported type/currency combinations?

piersy replied:
It should be yes

aaronmgdr replied:
the serializer only knows what it knows. If the transaction fit as the celo specific one then it delegates to the serializer for op chains which in turn delegates to the default serializer. We wouldnt want to block unknowns configurations as they could known by the other serializer. 

Our focus on viem was getting the cip64 type to work. not blocking unsupported transactions types. 


If you put "legacy" as the type if anything it will take that to mean eth type 0 transaction not the celo legacy. 


I think the use case you are building here is pretty unique, i doubt anyone else tries this. 


If you wish to test sending txns to the blockchain one option would be to create your own serializers 

```ts
function serializeBaddies(transaction) {
 if  (type == 'celo-legacy')  {
 const rlpEncode = toRlp([
    stringNumberToHex(transaction.nonce),
    transaction.gasPrice,
    transaction.gas,
    transaction.feeCurrency,
    transaction.gatewayFeeRecipient,
    transaction.gatewayFee,
    transaction.to,
    transaction.value,
    transaction.data,
    stringNumberToHex(transaction.chainId),
    '0x',
    '0x',
  ])
 } else if type === 'cip42' {
  ...
 
 }
 else
    return  serializeTransactionImporFromCelo(tx)
}
```

aaronmgdr replied:
then add that serializer to the chain def 

`const celoWithOldTxns = defineChain({...celo, serializer: serializeBaddies})`

piersy replied:
Thanks for the clarification @aaronmgdr!

piersy replied:
So rather than trying to bend viem into sending transactions it shouldn't I've added a golang test that sends unsupported transactions and checks that they are not accepted by the node.

ezdac reviewed (changes requested):
Looks good so far! I added some suggestions and questions regarding unsupported tx types.
This branch requires a rebase against `celo10`.

piersy committed:
Add smart contract interactions to smoke test

piersy committed:
Use writeContract instead of sendTransaction

piersy committed:
Remove trailing space

piersy committed:
Extract common viem setup code to separte file

Note that in many cases we were able to remove explicit use of account
in transaction creattion because it is "hoisted" in the walletClient.

Also removed chain-id and ACC_PRIVKEY as arguments from send_tx.mjs
because they are controlled by env vars, which are picked up by the mjs
file.

piersy committed:
Correct formatting

karlb commented on .github/workflows/ci.yaml:
```diff
@@ -41,10 +41,14 @@ jobs:
           shell: bash
           run: e2e_test/run_all_tests.sh
 
+        - name: Run e2e tests alfajores
+          shell: bash
+          run: NETWORK=alfajores e2e_test/run_all_tests.sh
+
```
Do we want to run on alfajores on every push?

piersy replied:
Probably not. I guess we want to run these on alfajores when alfajores is updated(not sure how we can know that) and when the e2e tests are updated(probably can do this with github actions)

karlb replied:
We could put it in a separate workflow and run it on a schedule (once a week?) and on request (`on: workflow_dispatch` puts a button into the GH actions web UI). But maybe just running it from a local checkout whenever we want is good enough.

piersy replied:
Oh nice, didn't know you could run workflows on a schedule, that sounds good to me, but I'd go for daily.

karlb commented:
Why is the CI workflow with the tests not run for the last commits? The workflow file looks fine to me.

piersy commented:
> Why is the CI workflow with the tests not run for the last commits? The workflow file looks fine to me.

Hey @karlb, no idea I only made one change to CI which was to run these tests on alfajores. Maybe because the automatic base change or the conflicts it now has?


Update: The workflow seemed to get run again after rebasing this on celo10, perhaps it was the existence of conflicts between this and celo10 that caused the workflow to not get run.

piersy committed:
Move assertions next to fetching actions

piersy committed:
Parameterise viem smoketests

piersy committed:
Fix just enough gas test

piersy committed:
Update skipped test to not reference missing function

aaronmgdr commented on e2e_test/js-tests/test_viem_smoketest.mjs:
```diff
@@ -0,0 +1,134 @@
+	return await walletClient.sendTransaction({
+		to: "0x00000000000000000000000000000000DeaDBeef",
+		value: 1,
+		type: type,
```
viem doesnt use this type field.

piersy replied:
Hey @aaronmgdr I think it must be using it because the tests verify that the transaction retrieved from the node has the same type passed to sendTransaction. How else could the types be matching up, see [here](https://github.com/celo-org/op-geth/pull/261/files#diff-729cf14983159ee568c6e3d85cc7d4b92cb7c7271fc160fe79dc7dfc4753e915R19)

aaronmgdr replied:
so kinda the opposite. 

when you deserialize / recover the transaction from the rlp it adds the type. 


I was slightly incorrect it does seem like viem will use the field to serialize if the type is there and known but if its not known it uses heuristics like does the transaction have gasPrice or does it have maxFeePerGas field to figure out type.

piersy replied:
Ok got you, thanks @aaronmgdr!

karlb commented on e2e_test/test_smoketest.sh:
```diff
@@ -0,0 +1,9 @@
+source shared.sh
+prepare_node
+
+(cd debug-fee-currency && forge build .)
```
This might solve the CI error:
```suggestion
(cd debug-fee-currency && forge build)
```

Alternatively, upgrading to the same foundry version we use in optimism now (143abd6a768eeb52a5785240b763d72a56987b4a) could help, since that version works for me locally even with out changes.

piersy committed:
Update e2e_test/test_smoketest.sh

Co-authored-by: Karl Bartel <karl.bartel@clabs.co>

piersy committed:
Move unsupported tx tests from viem to go

Due to the way that viem is structured it is a bad idea for us to block
transactions based on the supplied tx attributes, since they may be supported by
downstream components such as optimism's custom viem functionality.

So rather than trying to test unsupported txs via viem, we now test via
golang which allows us full control of the sent transaction, thereby
ensuring that rejection is happening at the node level.

This commit also extends the error message for deprecated transactions
to distinguish between celo legacy and plain legacy transactions.

piersy committed:
Don't run golang smoketest run with usual go tests

karlb commented on e2e_test/smoketest_unsupported_txs/unsupported_txs_test.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+func parsePrivateKey(privateKey string) (*ecdsa.PrivateKey, error) {
+	if len(privateKey) >= 2 && privateKey[0] == '0' && (privateKey[1] == 'x' || privateKey[1] == 'X') {
```
`if len(privateKey) == 66` would also have been an easy way to check for a `0x`-prefixed private key.

piersy committed:
Trigger e2e test runs on alfajores differently

Adds a new workflow that triggers e2e test runs on alfajores once a day,
and also in PRs whenever the e2e tests are updated.

piersy committed:
Add debug to see why file is not found

piersy committed:
Ensure tree is installed

piersy committed:
Update before install

piersy committed:
Update forge version and reinstate dot

piersy committed:
Provide explicit paths to forge

piersy committed:
Remove debug

piersy committed:
Merge branch 'celo10' into piersy/smoketest

piersy committed:
Merge branch 'celo10' into piersy/smoketest

ezdac commented on e2e_test/smoketest_unsupported_txs/unsupported_txs_test.go:
```diff
@@ -0,0 +1,105 @@
+	flag.StringVar(&feeCurrency, "fee-currency", "", "address of the fee currency to use")
+}
+
+func TestTxSendingFails(t *testing.T) {
```
Will this be discovered by a `go test ./...` and then lack the flags?
If so, maybe for now skip the test if the flags are not present and later we can spawn a local test environment if the flags are not present?

piersy replied:
No it has a build flag, so it's only picked up if you specify `-tags smoketest`

ezdac reviewed (approved):
Looks great now!

karlb commented:
Is there any reason for not having merged this yet, or is it just the merge conflicts?

piersy committed:
Merge remote-tracking branch 'origin/celo10' into piersy/smoketest

piersy commented:
> Is there any reason for not having merged this yet, or is it just the merge conflicts?

I think it just got superseded by our bad block problems, I've updated it now, so should be able to merge shortly 🤞

piersy committed:
Merge branch 'celo10' into piersy/smoketest

piersy committed:
Merge branch 'celo10' into piersy/smoketest
